subsys/bluetooth/services/ancs_client.c

```diff
@@ -368,15 +364,13 @@ static void bt_ancs_cp_write_callback(struct bt_conn *conn, uint8_t err,
 
 	atomic_clear_bit(&ancs_c->state, ANCS_CP_WRITE_PENDING);
 
-	struct bt_ancs_evt ancs_evt;
-
-	ancs_evt.evt_type = BT_ANCS_EVT_NP_ERROR;
-	ancs_evt.err_code_np = err;
-
-	ancs_c->evt_handler(&ancs_evt);
+	if (ancs_c->cp_write_cb) {
+		ancs_c->cp_write_cb(ancs_c, err);
+	}
 }
 
-int bt_ancs_cp_write(struct bt_ancs_client *ancs_c, uint16_t len)
+int bt_ancs_cp_write(struct bt_ancs_client *ancs_c, uint16_t len,
+		     bt_ancs_write_cb func)
```

Comment:
Not validating the func not NULL here, is this optional?

Reply:
It is possible that the callback function is NULL. e.g. the user does not check the GATT write response.

A NULL pointer check is done in bt_ancs_cp_write_callback().

A careless mistake is fixed in bt_ancs_cp_write_callback(). The callback function variable should be copied before clearing the atomic bit. It avoids a racing condition in a multi-threading system.